### PR TITLE
[Tiff] Improve performance, use common zlib class, cleanup

### PR DESCRIPTION
**Pull request by IldarKhayrutdinov**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

* Use common `zlib` class for tiff deflate compressor.
* Remove `SubStream` class.
* Improve performance of tiff colors.
* Cleanup.

Improvement report for `.NET Core 3.1`

|               TestImage            | *Ratio befor | *Ratio after | Imorivement |
|----------------------------------- |------------- |------------- |------------ |
|                medium_bw_Fax3.tiff |        7.29  |        7.31  |  -0.2%      |
|                 medium_bw_Rle.tiff |        7.32  |        7.48  |    -2%      |
| medium_grayscale_uncompressed.tiff |        1.60  |        1.15  |    28%      |
|   medium_palette_uncompressed.tiff |        2.88  |        2.94  |    -2%      |
|            medium_rgb_deflate.tiff |        2.60  |        1.58  |    39%      |
|                medium_rgb_lzw.tiff |        4.06  |        3.33  |    18%      |
|           medium_rgb_packbits.tiff |        1.01  |        0.45  |    55%      |
|       medium_rgb_uncompressed.tiff |        0.98  |        0.42  |    57%      |

*Ratio relatively System.Drawing

### CODE REVIEW

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero1TiffColor.cs [hunk 1]
```diff
@@ -25,6 +26,8 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
 
             int offset = 0;
 
+            var black = new Rgba32(0, 0, 0, 0xff);
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero1TiffColor.cs:29]: you can use named colors here instead `Color.White`  and `Color.Black`

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero8TiffColor.cs [hunk 2]
```diff
@@ -24,13 +25,16 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
 
             int offset = 0;
 
+            var rgba = default(Rgba32);
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero8TiffColor.cs:28]: You can use the `L8` data type directly instead or rgba. Its ment for gray image data

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgb888TiffColor.cs [hunk 3]
```diff
@@ -24,6 +25,7 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
 
             int offset = 0;
 
+            var rgba = default(Rgba32);
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgb888TiffColor.cs:28]: same here, use `L8`

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero1TiffColor.cs [hunk 4]
```diff
@@ -24,6 +25,8 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
 
             int offset = 0;
 
+            var black = new Rgba32(0, 0, 0, 0xff);
+            var white = new Rgba32(0xff, 0xff, 0xff, 0xff);
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero1TiffColor.cs:29]: use named colors

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero4TiffColor.cs [hunk 5]
```diff
@@ -25,27 +26,34 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
             int offset = 0;
             bool isOddWidth = (width & 1) == 1;
 
+            var rgba = default(Rgba32);
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero4TiffColor.cs:29]: use `L8` instead

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero8TiffColor.cs [hunk 6]
```diff
@@ -24,13 +25,16 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
 
             int offset = 0;
 
+            var rgba = default(Rgba32);
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero8TiffColor.cs:28]: use `L8`

File: src/ImageSharp/Formats/Tiff/Compression/TiffBaseCompression.cs [hunk 7]
```diff
@@ -55,9 +56,32 @@ public TiffBaseCompression(MemoryAllocator allocator, int width, int bitsPerPixe
...
-        /// <param name="stream">The <see cref="Stream"/> to read image data from.</param>
+        /// <param name="stream">The <see cref="Stream" /> to read image data from.</param>
+        /// <param name="stripOffset">The strip offset of stream.</param>
+        /// <param name="stripByteCount">The number of bytes to read from the input stream.</param>
+        /// <param name="buffer">The output buffer for uncompressed data.</param>
+        public void Decompress(BufferedReadStream stream, uint stripOffset, uint stripByteCount, Span<byte> buffer)
+        {
+            if (stripByteCount > int.MaxValue)
+            {
+                TiffThrowHelper.ThrowImageFormatException("Too big value of StripByteCount.");
```
Comment (brianpopow) [inline src/ImageSharp/Formats/Tiff/Compression/TiffBaseCompression.cs:67]:
```suggestion
                TiffThrowHelper.ThrowImageFormatException("The StripByteCount value is too big.");
```

**Review by brianpopow — CHANGES_REQUESTED** [review]
just a few small changes

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero1TiffColor.cs (hunk 1 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero1TiffColor.cs:29]: Oh, yeah, thanks 😌

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero8TiffColor.cs (hunk 2 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero8TiffColor.cs:28]: 👌

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgb888TiffColor.cs (hunk 3 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgb888TiffColor.cs:28]: it seems rgba is needed here, 3 components are used

Comment (brianpopow) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgb888TiffColor.cs:28]: yes you are right, sorry

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero1TiffColor.cs (hunk 1 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero1TiffColor.cs]: Done 798603f46cd92c21c4cf0aa74a2e2082b83d7ae7

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero8TiffColor.cs (hunk 2 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/BlackIsZero8TiffColor.cs]: Done 798603f46cd92c21c4cf0aa74a2e2082b83d7ae7

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero1TiffColor.cs (hunk 4 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero1TiffColor.cs]: Done 798603f46cd92c21c4cf0aa74a2e2082b83d7ae7

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero4TiffColor.cs (hunk 5 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero4TiffColor.cs]: Done 798603f46cd92c21c4cf0aa74a2e2082b83d7ae7

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero8TiffColor.cs (hunk 6 above)
Comment (IldarKhayrutdinov) [reply src/ImageSharp/Formats/Tiff/PhotometricInterpretation/WhiteIsZero8TiffColor.cs]: Done 798603f46cd92c21c4cf0aa74a2e2082b83d7ae7

**Review by brianpopow — APPROVED** [review]
спасибо!

**Comment by IldarKhayrutdinov** [comment]
Всегда пожалуйста! :)

**Pull request merged**